haystack/components/builders/hypothetical_doc_query.py

```diff
@@ -0,0 +1,127 @@
+        Question: {{question}}
+        Paragraph:
+        """
+        )
+        self.generator = OpenAIGenerator(
+            model=instruct_llm, generation_kwargs={"n": nr_completions, "temperature": 0.75, "max_tokens": 400}
+        )
+
+        self.pipeline = Pipeline()
+        self.pipeline.add_component(name="prompt_builder", instance=self.prompt_builder)
+        self.pipeline.add_component(name="generator", instance=self.generator)
+        self.pipeline.connect("prompt_builder", "generator")
+
+    def to_dict(self) -> Dict[str, Any]:
+        # ToDo
+        pass
+
+    def from_dict(cls, data: Dict[str, Any]) -> "InstructLM":
+        # ToDo
+        pass
+
+    @component.output_types(replies=List[str])
+    def run(self, query: str):
+        result = self.pipeline.run(data={"prompt_builder": {"question": query}})
+        return result["generator"]
+
+
+@component
+class HyDEBuilder:
+    """
+    Subcomponent to build the Hypothetical Document Embedding (HyDE) from the generated hypothetical documents.
+    """
+
+    def __init__(self):
+        self.embedder = SentenceTransformersDocumentEmbedder(
+            model="sentence-transformers/all-MiniLM-L6-v2", progress_bar=False
+        )
+        self.embedder.warm_up()
+
+    def to_dict(self) -> Dict[str, Any]:
+        # ToDo
+        pass
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "HyDEBuilder":
+        # ToDo
+        pass
+
+    @component.output_types(embedding=List[float])
+    def run(self, answers: List[str]):
+        embeddings = self.embedder.run([Document(content=answer) for answer in answers])
+        stacked_embeddings = array([doc.embedding for doc in embeddings["documents"]])
+        avg_embeddings = mean(stacked_embeddings, axis=0)
+        hyde_vector = avg_embeddings.reshape((1, len(avg_embeddings)))
+        return {"hypothetical_documents": answers, "hypothetical_embedding": hyde_vector[0].tolist()}
+
+
+@component
+class HypotheticalDocumentEmbedder:
+    """
+    Hypothetical Document Embeddings (HyDE)
+
+    Given a query, HyDE first zero-shot prompts an instruction-following language model to generate a "fake"
+    hypothetical document but that captures relevant textual patterns.
+
+    Then, it encodes the document into an embedding vector, identifying a neighborhood in the corpus
+    embedding space, from which similar real documents are retrieved based on vector similarity.
+
+    see:  "Precise Zero-Shot Dense Retrieval without Relevance Labels" (https://aclanthology.org/2023.acl-long.99/)
+
+    Usage example:
+    ```python
+    from haystack.components.embedders import HypotheticalDocumentEmbedder
+    query = "What should I see in the capital of France?"
+    hyde.run(query=query)
+    # [0.08679415807127952, -0.007719121221452951, 0.056458056718111035, -0.005992368655279278, ...]
+    ```
+    """
+
+    def __init__(self, instruct_llm: str = "gpt-3.5-turbo", nr_completions: int = 5):
```

Comment:
We should expose the api_key too as in the generator:
`api_key: Secret = Secret.from_env_var("OPENAI_API_KEY"),`
I understand that there are two models now but as only one can be customized here, wouldn't it make sense to call it just `model`?

Reply:
thanks - I just added the `api_key` as a parameter.

regarding the naming I followed the naming as in the paper, an instruct-following LLM to generate the fake documents.